### PR TITLE
Improve ArraySegment debugging

### LINKED ISSUE #88961: Improve ArraySegment debugging
ArraySegment could be improved with debugger display and debugger type proxy.

Current state:

image

See Span debugging support for ideas of improvement.

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #88961

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs [hunk 1]
```diff
@@ -276,6 +279,41 @@ private void ThrowInvalidOperationIfDefault()
             }
         }
 
+        /// <summary>
+        /// For <see cref="ArraySegment{Char}"/>, returns a new instance of string that represents the characters pointed to by the segment.
+        /// Otherwise, returns a <see cref="string"/> with the name of the type and the number of elements.
+        /// </summary>
+        public override string ToString()
```
Comment (huoyaoyuan) [inline src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:286]: Using `ToString` is an observable behavior change. The common approach is using a private property called `DebuggerDisplay`.

File: src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs [hunk 2]
```diff
@@ -276,6 +279,41 @@ private void ThrowInvalidOperationIfDefault()
...
+                StringBuilder stringBuilder = new StringBuilder(_count);
+
+                for (int i = 0; i < _count; i++)
+                {
+                    T element = this[i];
+                    stringBuilder.Append(Unsafe.As<T, char>(ref element));
+                }
+
+                return stringBuilder.ToString();
+            }
```
Comment (huoyaoyuan) [inline src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:299]: Don't use a `StringBuilder`, just call string constructor like `ReadOnlySpan<T>.ToString`.

File: src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs [hunk 3]
```diff
@@ -276,6 +279,29 @@ private void ThrowInvalidOperationIfDefault()
             }
         }
 
+        private string DebuggerToString()
+        {
+            if (typeof(T) == typeof(char))
+            {
+                return new string(new ReadOnlySpan<char>(ref Unsafe.As<T, char>(ref _array![_offset]), _count));
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:286]:
```suggestion
                return new ReadOnlySpan<char>(_array, _offset, _count).ToString();
```

File: src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs [hunk 4]
```diff
@@ -276,6 +279,29 @@ private void ThrowInvalidOperationIfDefault()
...
+            private readonly U[] _array;
+
+            public ArraySegmentDebugView(ArraySegment<U> arraySegments)
+            {
+                _array = arraySegments.ToArray();
+            }
+
+            [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
+            public U[] Items => _array;
+        }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:303]:
```suggestion
        private sealed class ArraySegmentDebugView(ArraySegment<T> array)
        {
            [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
            public T[] Items => array.AsSpan().ToArray();
        }
```

Comment (huoyaoyuan) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs]: Have we started to roll-in primary constructor in runtime?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs]: Yes, for simple uses like debugger type proxies.

File: src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs [hunk 5]
```diff
@@ -276,6 +279,22 @@ private void ThrowInvalidOperationIfDefault()
...
+            }
+
+            return $"System.ArraySegment<{typeof(T).Name}>[{_count}]";
+        }
+
+        private sealed class ArraySegmentDebugView<U>(ArraySegment<U> array)
+        {
+            [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
+            public U[] Items => array.AsSpan().ToArray();
+        }
```
[the remark this answers is not in the document]
Comment (pedrobsaila) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:296]: It fails on compilation CS0416 error : 'type parameter': an attribute argument cannot use type parameters

[the remark this answers is not in the document]
Comment (pedrobsaila) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs]: thanks for the help ^^

File: src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs [hunk 6]
```diff
@@ -276,6 +279,22 @@ private void ThrowInvalidOperationIfDefault()
...
         }
 
+        private string DebuggerToString()
+        {
+            if (typeof(T) == typeof(char))
+            {
+                return this.AsSpan().ToString();
+            }
+
+            return $"System.ArraySegment<{typeof(T).Name}>[{_count}]";
```
[the remark this answers is not in the document]
Comment (pedrobsaila) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:289]: I'm not sure to understand. What is the issue exactly with this implementation ?

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:289]: > The original issue wanted the elements to be displayed in such case I think.

We don't do that for other collections. I'm not sure why we'd do it here.

[the remark this answers is not in the document]
Comment (pedrobsaila) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:289]: Can someone explains to me what should be done here ? I used the same code as the one for Span

https://github.com/dotnet/runtime/blob/ac02b66e41041e1cefd355eacaddb45bcb4f1741/src/libraries/System.Private.CoreLib/src/System/Span.cs#L376-L387

[the remark this answers is not in the document]
Comment (JamesNK) [reply src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:289]: For types with a `Count` such as this, the debugger display usually displays `Count = XXX`.

e.g. https://github.com/dotnet/runtime/blob/f9b07d43eab8f3ba46ecc6778d8d1d85563362cd/src/libraries/System.ObjectModel/src/System/Collections/ObjectModel/KeyedCollection.cs#L12

File: src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs [hunk 7]
```diff
@@ -19,6 +20,8 @@ namespace System
     // (ie, users could assign a new value to the old location).
     [Serializable]
     [TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
+    [DebuggerDisplay("{DebuggerToString(),nq}")]
+    [DebuggerTypeProxy(typeof(ArraySegment<>.ArraySegmentDebugView))]
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/ArraySegment.cs:24]: I believe that we should simply mimic what `List<T>` does:

https://github.com/dotnet/runtime/blob/568ed4125d0ae62c0d82d8f3703854fcc04bf32d/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L17C1-L18

`ArraySegment<T>` implements `IList<T>` which extends `ICollection<T>` so it should work OOTB without any other code changes being required


```suggestion
    [DebuggerTypeProxy(typeof(ICollectionDebugView<>))]
    [DebuggerDisplay("Count = {Count}")]
```

**Review by adamsitnik — CHANGES_REQUESTED** [review]
@pedrobsaila thank you for your contribution!

I think that we can simplify the solution, PTAL at my suggestion. Thank you!

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @pedrobsaila !

**Pull request merged**